NineChronicles.Headless/ActionEvaluationPublisher.cs

```diff
@@ -84,59 +78,169 @@ public async Task AddClient(Address clientAddress)
                 MaxReceiveMessageSize = null
             };
 
-            var channel = GrpcChannel.ForAddress($"http://{_host}:{_port}", options);
-            var client = await StreamingHubClient.ConnectAsync<IActionEvaluationHub, IActionEvaluationHubReceiver>(
-                channel,
-                null!
-            );
-            await client.JoinAsync(clientAddress.ToHex());
-            if (_clients.TryAdd(clientAddress, (client, ImmutableHashSet<Address>.Empty)))
+            GrpcChannel channel = GrpcChannel.ForAddress($"http://{_host}:{_port}", options);
+            Client client = await Client.CreateAsync(channel, clientAddress, _context);
+            if (_clients.TryAdd(clientAddress, client))
             {
                 if (clientAddress == default)
                 {
                     Log.Warning("[{ClientAddress}] AddClient set default address", clientAddress);
                 }
 
                 Log.Information("[{ClientAddress}] AddClient", clientAddress);
+                client.Subscribe(
+                    _blockRenderer,
+                    _actionRenderer,
+                    _exceptionRenderer,
+                    _nodeStatusRenderer
+                );
+            }
+            else
+            {
+                await client.DisposeAsync();
             }
+        }
 
-            if (!_blockRenderDictionary.ContainsKey(clientAddress))
+        public override async Task StopAsync(CancellationToken cancellationToken)
+        {
+            foreach (Client? client in _clients.Values)
             {
-                _blockRenderDictionary[clientAddress] = true;
-                _blockRenderer.BlockSubject.SubscribeOn(NewThreadScheduler.Default).Subscribe(
-                    async pair =>
-                    {
-                        try
+                if (client is { })
+                {
+                    await client.DisposeAsync();
+                }
+            }
+            await base.StopAsync(cancellationToken);
+        }
+
+        public void UpdateSubscribeAddresses(byte[] addressBytes, IEnumerable<byte[]> addressesBytes)
+        {
+            var address = new Address(addressBytes);
+            if (address == default)
+            {
+                Log.Warning("[{ClientAddress}] UpdateSubscribeAddresses set default address", address);
+            }
+            var addresses = addressesBytes.Select(a => new Address(a)).ToImmutableHashSet();
+            if (_clients.TryGetValue(address, out Client? client) && client is { })
+            {
+                client.TargetAddresses = addresses;
```

Comment:
is this ok not use `TryUpdate`?

Reply:
Basically, all `client` in `_clients` are ref types so it will affect objects in the collections without explicit updating. but... yeah, we may need to consider sync behavior.

Reply:
I've added https://github.com/planetarium/NineChronicles.Headless/pull/1682/commits/e42b60053092abce9a239241f8ea237785a7b640 to address it. 

Also, I turned this PR into a draft until testing. cc @ipdae @riemannulus @moreal

Reply:
Confirm.

Reply:
Amended. https://github.com/planetarium/NineChronicles.Headless/pull/1682/commits/cf210a1a3486cb666368b6a333b32c2e2044686e cc @riemannulus @moreal @ipdae